src/ucp/api/ucp.h

```diff
@@ -3135,6 +3108,169 @@ ucs_status_ptr_t ucp_worker_flush_nb(ucp_worker_h worker, unsigned flags,
+ *
+ * The structure defines the parameters that are used for the
+ * UCP active message function
+ */
+typedef struct ucp_am_params {
+  uint64_t field_mask ;
+  uint64_t flags ;
+  /* Number of elements that the ucp library should allocate in the 'iovec'
+   * for the callback to fill in.
+   */
+  size_t iovec_size ;
+} ucp_am_params_t ;
+
+/*
+ * The ucp library will drive a function on the arrival of each fragment
+ * of data from the initiator.
+ */
+typedef void (*ucp_am_data_function_t)(
+                                       void *target,
+                                       void *source,
+                                       size_t bytes,
+                                       void *cookie
+                                      ) ;
+/*
+ * The ucp library will drive a function when the transfer from the
+ * initiator is complete.
+ */
+typedef ucs_status_t (*ucp_am_local_function_t)(
+                                                void *arg,
+                                                void *cookie,
+                                                ucp_dt_iov_t *iovec,
+                                                size_t iovec_length
+                                               ) ;
+
+/*
+ * Structure for communicatiob between the active message
+ * callback and the ucp library. The ucp library sets up 'iovec_max_length'
+ * to indicate the length of the iovec in this structure; the remaining
+ * fields are set by the callback to indicate what should happen
+ * with the remaining data from the initiator. In the initial implementation,
+ * iovec_max_length is always 1.
+ */
+typedef struct ucp_am_recv {
+    /* Function to be driven when data transfer is complete */
+    ucp_am_local_function_t local_fn;
+    /* Argument to be passed to local_fn */
+    void                   *cookie;
+    /* Function to be driven when each fragment of data transfer is complete
+     * In the initial implementation, all data is transferred in one fragment,
+     * so the data_fn isriven once just before the local_fn .
+     * This implementation is to ease porting of applicaitons currently
+     * coded to use IBM PAMI.
+     */
+    ucp_am_data_function_t  data_fn;
+    /* Argument to be passed to data_fn */
+    void                   *data_cookie ;
+    /* Sise of iovec allocated by the ucp library */
+    size_t                  iovec_max_length;
+    /* Size of iovec filled in by the callback */
+    size_t                  iovec_length;
+    /* iovec indicating where the data from the initiator is to be placed */
+    ucp_dt_iov_t            iovec[1];
+  } ucp_am_recv_t;
+
+  /**
+   * @ingroup UCP_ENDPOINT
+   * @brief Callback to process incoming Active Message.
+   *
+   * If the UCP_AM_SEND_RENDEZVOUS flag is set, and the AM is for an iovec with
+   * exactly 2 elements, and the first element length does not exceed a
+   * threshold (32 bytes in the initial implementation), then the AM gets
```

Comment:
what does this 32-byte thresholds define? Why need to check it?

Reply:
This is the largest size of iovec[0] that is supported by the AM Rendezvous implementation. There is a fixed-size array in the initial packet from the initiator to the target, currently sized at 32 bytes; making it larger would waste IB bandwidth. 32 bytes is sufficient for my purposes; someone else might want to make it larger. If iovec[0] is larger than this, the ucp library will carry the AM in then non-RDMA way.